### ISSUE TITLE
`java.lang.Process.waitFor` unable to handle broken pipe

### ISSUE THREAD

**Issue by WojciechMazur** [scala-native/scala-native#3944]
Based on failures in https://github.com/com-lihaoyi/os-lib/pull/274
Waiting for a process that is terminated due to a broken pipe is impossible to catch - `waitFor` returns false even though the process has been terminated already. Underlying `kevent`/`ppoll` would fail with `Invalid argument` errno

A self-contained reproducer would be useful to debug the issue - os-lib does spin up an additional threads delegated to pumping process stdin/stdout

Reproducer 
```scala
//> using dep com.lihaoyi::os-lib::0.10.2

import java.nio.file.*
import java.io.IOException

@main def Test = {
  val writerSourceCode = """
    |import scala.util.Try
    |
    |object TestJarWriter {
    |  def main(args: Array[String]): Unit = {
    |    val writeN = Try(args(0).toInt).getOrElse(-1)
    |    val writeSleep = Try(args(1).toInt).getOrElse(0)
    |    val debugOutput = Try(args(2).toBoolean).getOrElse(false)
    |    val stdin = new java.util.Scanner(System.in)
    |    
    |    var i = 0
    |    while (writeN == -1 || i < writeN) {
    |      println(s"Hello $i")
    |      if (debugOutput) Console.err.println(s"Written $i")
    |      Thread.sleep(writeSleep)
    |      i += 1
    |    }
    |    stdin.close()
    |    if (debugOutput) Console.err.println("Exiting writer")
    |  }
    |}""".stripMargin

  val readerSourceCode = """
  |import scala.io.StdIn
  |import scala.util.Try
  |
  |object TestReader {
  |  def main(args: Array[String]): Unit = {
  |    val readN = Try(args(0).toInt).getOrElse(-1)
  |    val readSleep = Try(args(1).toInt).getOrElse(0)
  |    val debugOutput = Try(args(2).toBoolean).getOrElse(false)
  |    val stdin = new java.util.Scanner(System.in)
  |
  |    var i = 0
  |    while (readN == -1 || i < readN) {
  |      if (debugOutput) Console.err.println(s"At: $i")
  |      val read = stdin.nextLine()
  |      if (read != null) println(s"Read: $read")
  |      else println("Read: (EOF reached)")
  |      Thread.sleep(readSleep)
  |      i += 1
  |    }
  |    stdin.close()
  |    if (debugOutput) Console.err.println("Exiting reader")
  |  }
  |}
  """.stripMargin

  def runScalaCode(code: String, args: String*): Seq[String] = {
    val sourceFile =
      Files.write(Files.createTempFile("", ".scala"), code.getBytes())
    Seq("scala", sourceFile.toAbsolutePath().toString()) ++ args
  }

  def writerProc(
      count: Int,
      sleep: Int,
      debugOutput: Boolean = true
  ): Seq[String] =
    runScalaCode(
      writerSourceCode,
      count.toString,
      sleep.toString,
      debugOutput.toString()
    )
  def readerProc(
      count: Int,
      sleep: Int,
      debugOutput: Boolean = true
  ): Seq[String] =
    runScalaCode(
      readerSourceCode,
      count.toString,
      sleep.toString,
      debugOutput.toString()
    )

  locally {
    val p = os
      .proc(writerProc(-1, 0, false))
      .pipeTo(os.proc(readerProc(3, 0, true)))
      .spawn()      
    assert(true == p.waitFor(10000), "waitFor")
    p.destroy()
    assert(137 == p.exitCode(), "exitCode")
  }

  System.exit(0) // os-lib spins a non-deamon thread
}

```

**Comment by LeeTibbert** [comment]
Debug Notes & early morning ramblings:

I will probably start first with the Linux/macOS code where I have a better understanding
of the API and can build things but I would like to capture this information.

Based on the bird-in-hand error in a Windows test case for PR #3949. 

```
 override def waitFor(timeout: scala.Long, unit: TimeUnit): scala.Boolean =
    synchronized {
      import SynchApiExt._
      def hasValidTimeout = timeout > 0L
      def hasFinished =
        WaitForSingleObject(
          handle,
          unit.toMillis(timeout).toUInt
	) match {
          case WAIT_TIMEOUT => false
          case WAIT_FAILED =>
            throw WindowsException("Failed to wait on proces handle")
          case _ => true
	}

      !isAlive() ||
        (hasValidTimeout && hasFinished)
    }
```

To examine the code, given the information of recent failures, not to pick on it.

* I believe there are both race conditions and extra work being done. 
  `!isAlive() || (hasValidTimeout && hasFinished)`. Just because the process
   isAlive at the beginning of the line, does not mean that it isAlive when `WaitForSingleObject()` is
   called.

  *  hasValidTimeout should probably be done at method entry, not in an OR. It probably  
     wants to throw it own InvalidArgument Exception, specifying that it is complaining
     about a negative timeout.

* The best solution for avoiding race conditions at this level is probably to rely on the OS.

  `WaitForSingleObject()` has the definitive and most atomic answer for 
   whether the handle is open.  I think a solution will involve handling any "invalid argument"
   error in the catch block.  
  
I think that checking for valid handle at method entry will prove to be unnecessary and cycle-consuming.
The race condition will still exist. It might be make error reporting at method entry   easier.  It may also
make InvalidArgument Exception recovery simpler if that code knows that the handle was valid
at method entry.    I'd have to write the code to figure that out.  For now, I think any check at method
entry should be DEBUG_ONLY and the (short) method body should do optimistic "expect success" & 
recover  in the "catch" if `WaitForSingleObject` objects. The means branching in the "catch" on "Invalid Handle".
(and probably punting the question "Was it garbage to begin with, or did we get called with a closed handle
or did someone asynchronously close the handle on us?"
 
```
From URL: https://stackoverflow.com/questions/5603625/how-to-check-if-a-handle-is-valid-or-not
`GetHandleInformation` or 
```
Not quite true. When the handle is a pipe, it can be closed by the other party. (and is invalidated by the other party). Calling PeekNamedPipe() you will get an error return value, and the error is invalid handle when you call getLastError().
```

**Comment by LeeTibbert** [comment]
Debug note:  Reproduced on macOS (14.mumble, M1) via:
```
$ scala-cli --native --native-version 0.5.3 I3944_Reproducer.scala
Exception in thread "main"java.io.IOException: wait pid=15408, kevent failed: Invalid argument
	at java.lang.process.UnixProcessGen2.bsdWaitForImpl(Unknown Source)
	at java.lang.process.UnixProcessGen2.osWaitForImpl(Unknown Source)
	at java.lang.process.UnixProcessGen2.waitWithRepeat$1(Unknown Source)
	at java.lang.process.UnixProcessGen2.waitFor$$anonfun$1(Unknown Source)
	at java.lang.process.UnixProcessGen2$$Lambda$7.apply(Unknown Source)
	at scala.Option.orElse(Unknown Source)
	at java.lang.process.UnixProcessGen2.waitFor(Unknown Source)
	at os.SubProcess.waitFor(Unknown Source)
	at os.ProcessPipeline.waitForRec$1(Unknown Source)
	at os.ProcessPipeline.waitFor(Unknown Source)
	at I3944_Reproducer$package$.Test(Unknown Source)
	at Test.main(Unknown Source)
	at <none>.main(Unknown Source)

```
WAG: Looks like Option `_exitValue` is not getting set when it should have been. Now Why? 

Next step, I need to be able to run against a debugging javalib to understand what is going on.
Even if `_exitValue` is not getting set, the pid should still be valid (i.e. not exited & "waited" on).

**Comment by LeeTibbert** [comment]
FYI, for those following along from home, Work In Progress report:

1) I have the reproducer from the base topic running on a private copy of Scala Native 0.5.4-SNAPSHOT:
    `$ scala-cli --native --native-version 0.5.4-SNAPSHOT --scala-version 3 I3944_Reproducer_LeeT.scala`

2) `ppoll` & `kexec` no longer complain (using the proces.isAlive() correction in mainline).

3) The reproducer has an assertion expecting 138 (SIGKILL + 128) after the p.destroy().  I think 
     os-lib has `destroyForcibly` semantics for `destroy()` of an os-lib process or pipeline. I will have
     to do more tracing.  138 seems weird, but that is what the reproducer expects.

     Scala Native is returning 127 (process exited or not found) for processes which have indeed exited.
      Sending them a SIGTERM or SIGKILL after  successful exit should probably not change that status.

       I will have to trace what JVM does (with destroyForcibly() after a successful termination.  Problem
       for another sprint.

4)  No data seems to be being sent across the pipeline (or it is sent and Reader can not read it).  
     `Problem` for a week of sprints. I want to get the lower levels, such as expected exitValue working
     before I loose that much time: build a firm foundation first.

At least I am moving now.  Life is good (or at least getting better).

**Comment by LeeTibbert** [comment]
I believe that I have discovered a problem and have a trial fix.
With it, and the `isAlive` fix currently in mainline, the reproducer in the 
base topic passes (on Linux, have not tried macOS yet).

Appears that do to some bogus math, a time value was going negative. `ppoll()` & `kexec()` us all along
that there was an invalid argument.  Just have to listen hard enough to what they were telling to tell us.

I hope to create a PR next activity cycle.

**Comment by LeeTibbert** [comment]
PR #3954 passes CI, all green.  That, together with the now merged `isAlive` PR fix the problem described
in the base topic.

Yes, the relationship between the fix and the cause is not face evident.

More description later, when the day cools off enough for the next sprint.

**Comment by LeeTibbert** [comment]
The `errno 127` problem I described above was due to the fact that I did not have Scala directly installed on
that system.  

`ppoll` was doing its job. I should have listened to it or understood it better. Errno 127 generally means
" 'Command not found" and that was indeed the case.

**Comment by LeeTibbert** [comment]
##### Closing summary

The key to understanding how PR #3954 fixes this Issue is in the error message.
In the os-lib failing tests, it is given as:
`Exception in thread "main"java.io.IOException: wait pid=15408, ppoll failed: errno 22`
(yeah "errno 22", that is helpful! but I should be grateful for every hint I get and can Google errno 22)

When the reproducer above runs successfully on macOS, it gives the equivalent but more
informative:
`Exception in thread "main"java.io.IOException: wait pid=15408, kevent failed: Invalid argument` 

Now which argument was invalid?  Looking at the arguments common across the two calls, pid (process id)
and timeout (struct timespec) are the likely candidates.  

unix-like `struct timespec` has two Long fields `tv_sec` and `tv_nsec`.  `tv_nsec` is required to be in the 
closed interval [0 to 999,999,999]. If not, macOS `kevent` and unix-like `ppoll` set errno to EINVAL (22)
and return -1.

Examining the math of the "couldn't possibly be wrong" `fillTimeval()` (now `fillTimespec()` given the
Bayesian information that "something is invalid somewhere" revealed several errors, fixed in PR #3945.

`TimeUnit`s larger than `TimeUnit.SECOND` always have zero nanoseconds.

Now consider passing in Long.MAX_VALUE of `TimeUnit.MILLISECONDS`. This should set the `tv_sec`
field with the number of full seconds and the `tv_nsec` field with the remaining number of nanoseconds less than
a full second, if any.

To trace a closely relevant example, use the "10 * 1000" `TimeUnit.MILLISECOND` values from the reproducer 
` assert(true == p.waitFor(10000), "waitFor")`.

The prior math would calculate the `tv_sec` field correctly. 

To set the `tv_nsec` field, it would convert the full 10 seconds to (10 * 1_000_000_000) nanoseconds and
then subtract off 2 milliseconds converted to nanoseconds (2 * 1_000_000), leaving a number much larger
then 999_999_999. The exact large value is left as an exercise for the reader, but is essentially 10 million million.

`ppoll` and `kevent` had every right and a duty to report EINVAL.
 

PR #3954 corrects this math.  

It also accounts for saturation or overflow of the range of a `Long`.


That is, the combination `Long.MAX_VALUE` & `TimeUnit.MILLISECONDS` should give a
`tv_nsec` of `775807`.

The usual algorithm of "convert supplied timeout to nanoseconds and then subtract off whole seconds converted 
to nanoseconds" will fail because:

1)  TimeUnit.MILLISECONDS.toNanos(Long.MaxValue)` will saturate/overflow/clamp to Long.MaxValue
```
scala> val t_1 = TimeUnit.MILLISECONDS.toNanos(Long.MaxValue)
val t_1: Long = 9223372036854775807

```
2) 
```
scala> val t_2 = TimeUnit.MILLISECONDS.toSeconds(Long.MaxValue)
val t_2: Long = 9223372036854775
``` 
3) Number of seconds converted to nanos will saturate/overflow/clamp at Long.MAX_VALUE
```
scala> val t_3 = TimeUnit.SECONDS.toNanos(t_2)
val t_3: Long = 9223372036854775807

```
4)  `t_1` - `t_3` is zero, incorrectly setting `tv_nsec`.

5) Using modular arithmetic, as in the PR will correctly set `tv_nsec` to non-zero and strictly less than
     1_000_000_000 nanos
```
val t_4 = TimeUnit.MILLISECONDS.toNanos(Long.MaxValue % 1000)
val t_4: Long = 807000000

```
##### Epilogue

If you have read and traced this far and still have your wits & good humor, then you deserve a 
beverage or comfort food of your choice or personally relevant equivalent.
